### PR TITLE
Convert `Absinthe.Language.*` structs to graphql representation

### PR DESCRIPTION
**Pull request by maartenvanvliet**
This PR adds the inspect protocol for the Absinthe.Language.Document struct. 

A parsed graphql document can now be converted back to its graphql string representation. This gives the possible for a few new features:

  * In Elixir 1.13 formatting plugins can be used to format files or sigil delimited texts. This means .graphql documents can be formatted. See `Absinthe.Formatter` for an example.
  * Users could log incoming documents in a custom phase with interpolated arguments masked. (See https://github.com/absinthe-graphql/absinthe/issues/723). 
  * Programmatically create Absinthe.Language.Documents and convert them to graphql documents. 
  * Some instrumentation requires a canonical representation of a document. An incoming Absinthe.Language.Document can now be changed to become canonical (See https://www.apollographql.com/docs/studio/setup-analytics/#operation-signing)

### CODE REVIEW

File: lib/absinthe/formatter.ex [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        # Define the desired plugins
+        plugins: [Absinthe.Formatter],
+        # Remember to update the inputs list to include the new extensions
+        inputs: ["{mix,.formatter}.exs", "{config,lib,test}/**/*.{ex,exs}", "{lib, priv}/**/*.{gql,graphql}"]
+      ]
+
+  """
+
+  def features(_opts) do
+    [sigils: [], extensions: [".graphql", ".gql"]]
```
Comment (binaryseed) [inline lib/absinthe/formatter.ex:27]: Is there a use case for a `:G` sigil here?

File: lib/absinthe/language/render.ex [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+  end
+
+  def inspect(term, options) do
+    Inspect.Any.inspect(term, options)
+  end
+
+  defp render(bp)
+
+  defp render(%Absinthe.Language.Document{} = doc) do
+    doc.definitions |> Enum.map(&render/1) |> join("\n")
```
Comment (binaryseed) [inline lib/absinthe/language/render.ex:23]: I think I used `join(line())` in the other render module..

**Review by binaryseed — COMMENTED** [review]
This is looking really great!

Would you mind reviewing the test cases in this repo:
* https://github.com/newrelic/absinthe-schema-stitching-example/blob/master/apps/schema_stitch/test/query_generator_test.exs

At New Relic, we built a query renderer but it didn't use the Algebra. We did however, run into a variety of edge cases over time and fixed them each alongside tests. If all those test queries render nicely, I think we'll be good to go on this!

File: lib/absinthe/formatter.ex (hunk 1 above)
Comment (maartenvanvliet) [reply lib/absinthe/formatter.ex:27]: Since Absinthe does not define a sigil of its own I've left it out here. It's fairly simple for users to define their own formatter plugin and include any custom sigils.

File: lib/absinthe/language/render.ex (hunk 2 above)
Comment (maartenvanvliet) [reply lib/absinthe/language/render.ex:23]: Yes, in fact it uses `join([line(), line()])`. I'll change it to be consistent with the other module.

**Comment by maartenvanvliet** [comment]
I've run the documents in https://github.com/newrelic/absinthe-schema-stitching-example/blob/master/apps/schema_stitch/test/query_generator_test.exs and they render nicely :) 

The absinthe parser also accepts SDL. Rendering the Language.* structs for SDL however is not implemented yet. For completeness I'll see if I can do that as a followup.

**Comment by opsb** [comment]
This looks super useful. We've been wanting to sanitise queries for logging so I started down the road of creating a renderer for the blueprint structs. My plan is to allow fields to be whitelisted for logging in the schema e.g.

```
input_object :update_user_input do
  field :id, non_null(:id)
  field :email, :string, meta: [log: true]
  field :first_name, :string
  field :last_name, :string
end
``` 

The reason I'd thought to use the blueprint structs rather than the language structs is that they have the schema_nodes populated so it's possible to query the `meta` and determine if a field should be included in the rendered output or not. It seems excessive to have renderers both for the blueprint and language structs so I'm now pondering what the best path forwards is. Given this renderer exists, what's the best approach for log masking?

**Comment by benwilson512** [comment]
> We've been wanting to sanitise queries for logging so I started down the road of creating a renderer for the blueprint structs.

Can you elaborate on this? Sensitive parameters from end users should always be passed in as variables, and this already supports sanitization.

**Comment by opsb** [comment]
Hey @benwilson512, our absinthe API is client facing so we're not building the queries ourselves. Our position is that the spec supports inline arguments and variables so we should support them too. We do encourage using variables but for instance when exploring our API for the first time it's nicer to be able to just using arguments inline.

**Comment by benwilson512** [comment]
>  Our position is that the spec supports inline arguments and variables so we should support them too

That's fair, but understand that there are reasonable limits to what this sport entails. For example, it is _much_ easier to have syntax errors in the GraphQL than in JSON, and if the GraphQL has syntax errors then it can't easily be stripped of sensitive information. This is particularly complex because Absinthe.Plug logs values _prior_ to parsing, so this is not a simple change.

@opsb I think this question should be followed up in an issue on Absinthe.Plug after this PR is merged, since I think it is largely orthogonal to getting this PR over the finish line.

**Comment by opsb** [comment]
@benwilson512 that's an interesting point about syntax errors - possibly we'd just not log the query at all in that case. I think we'd probably skip the built in logger and introduce our own logging phase further on. But yeah, agree it make sense to explore this more in a follow on ticket.

**Review by binaryseed — APPROVED** [review]

**Comment by binaryseed** [comment]
@benwilson512 think this is ready for merge & release?

**Comment by benwilson512** [comment]
❤️ thank you @maartenvanvliet !
